crates/pglt_workspace/src/workspace/server/change.rs

```diff
@@ -69,37 +75,55 @@ impl Document {
         changes
     }
 
-    /// Applies a full change to the document and returns the affected statements
-    fn apply_full_change(&mut self, text: &str) -> Vec<StatementChange> {
+    /// Helper method to drain all positions and return them as deleted statements
+    fn drain_positions(&mut self) -> Vec<StatementChange> {
+        self.positions
+            .drain(..)
+            .map(|(id, _)| {
+                StatementChange::Deleted(Statement {
+                    id,
+                    path: self.path.clone(),
+                })
+            })
+            .collect()
+    }
+
+    /// Applies a change to the document and returns the affected statements
+    ///
+    /// Will always assume its a full change and reparse the whole document
+    fn apply_full_change(&mut self, change: &ChangeParams) -> Vec<StatementChange> {
         let mut changes = Vec::new();
 
-        changes.extend(self.positions.drain(..).map(|(id, _)| {
-            StatementChange::Deleted(Statement {
-                id,
-                path: self.path.clone(),
-            })
-        }));
+        changes.extend(self.drain_positions());
 
-        self.content = text.to_string();
+        self.content = change.apply_to_text(&self.content);
 
-        changes.extend(
-            pglt_statement_splitter::split(&self.content)
-                .ranges
-                .into_iter()
-                .map(|range| {
-                    let id = self.id_generator.next();
-                    let text = self.content[range].to_string();
-                    self.positions.push((id, range));
+        let (ranges, diagnostics) = document::split_with_diagnostics(&self.content, None);
 
-                    StatementChange::Added(AddedStatement {
-                        stmt: Statement {
-                            path: self.path.clone(),
-                            id,
-                        },
-                        text,
-                    })
-                }),
-        );
+        self.diagnostics = diagnostics;
+
+        // Do not add any statements if there is a fatal error
+        if self
+            .diagnostics
+            .iter()
+            .any(|d| d.severity() == pglt_diagnostics::Severity::Fatal)
```

Comment:
hard to see in the github diff, but is `self` not a `Document` and we can check the fatal_error helper?